### PR TITLE
lib: modem: lte_net_if: Fix handling of missing IPv4

### PR DESCRIPTION
**Pull request by glarsennordic**
Currently, if IPv4 is disabled or was not assigned, the cellular network is incorrectly treated as
offline, even if an IPv6 gets assigned.

This patch fixes this behavior.

### CODE REVIEW

**Comment by glarsennordic** [comment]
I need to do some additional testing of this, but I am pretty sure this will work since, once again, it is a simple change.

File: lib/nrf_modem_lib/lte_net_if/lte_net_if.c [hunk 1]
```diff
@@ -195,21 +195,20 @@ static void on_pdn_activated(void)
 #if CONFIG_NET_IPV4
 	int ret;
 
+	/* If IPv4 is enabled, check whether modem has been assigned an IPv4. */
 	ret = lte_ipv4_addr_add(iface_bound);
+
```
Comment (plskeggs) [inline lib/nrf_modem_lib/lte_net_if/lte_net_if.c:200]: Should there be a similar warning here `#if CONFIG_NET_IPV6` and `lte_ipv6_addr_add()` fails with `ENODATA` (assuming there is such a function and response)?

**Review by plskeggs — APPROVED** [review]

File: lib/nrf_modem_lib/lte_net_if/lte_net_if.c (hunk 1 above)
Comment (glarsennordic) [reply lib/nrf_modem_lib/lte_net_if/lte_net_if.c:200]: Interestingly, nope -- a separate event alerts to IPv6 assignment

Comment (eivindj-nordic) [reply lib/nrf_modem_lib/lte_net_if/lte_net_if.c:200]: I'm wondering, with `PDN_EVENT_IPV6_UP` as a separate event that calls `lte_ipv6_addr_add()` and `update_has_pdn(true)`, why is the `update_has_pdn(true)` needed here for IPv6? Shouldn't it be set to online when we get the  `PDN_EVENT_IPV6_UP` event?

Comment (glarsennordic) [reply lib/nrf_modem_lib/lte_net_if/lte_net_if.c:200]: Ah, you are right,  I do see now that `on_pdn_ipv6_up` calls `update_has_pdn(true)`

But I am surprised to see that. I thought `PDN_EVENT_IPV6_UP`/`PDN_EVENT_IPV6_DOWN` can only be relied upon to detect IPv6 availability

While IPv6 being UP does guarantee PDN is up, it is possible for IPv6 to go down without the PDN going down -- so PDN down must be detected with `PDN_EVENT_DEACTIVATED`. Therefore, it stands to reason (to my mind) that PDN up should be detected only based on `PDN_EVENT_ACTIVATED`.

In other words, I _**think**_ we should remove the `update_has_pdn(true)` from `on_pdn_ipv6_up`

The problem is, the exact conditions under which these PDN events are fired is not well documented, and I don't have a good way to properly test the various edge-cases (IPv6-only, IPv4-only, IPv6 gain -> IPv4 gain -> IPv6 loss, etc), so I don't have a way to verify that my suggestion here is definitely correct.

I see that you are the one who initially added the `update_has_pdn(true)` call to `on_pdn_ipv6_up` -- is my argument here convincing to you?

Comment (eivindj-nordic) [reply lib/nrf_modem_lib/lte_net_if/lte_net_if.c:200]: Based on the description in the AT commands guide and what I can see from the PDN sample with some modifications I agree. We get the `PDN_EVENT_ACTIVATED` event with PDP type IPV6 upon activation, and `PDN_EVENT_DEACTIVATED` upon deactivation.

Comment (glarsennordic) [reply lib/nrf_modem_lib/lte_net_if/lte_net_if.c:200]: Thanks for verifying it with some direct testing -- I'll go ahead and make that change, then!

File: doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst [hunk 2]
```diff
@@ -451,6 +451,7 @@ Modem libraries
 * :ref:`nrf_modem_lib_lte_net_if` library:
 
   * Added a log warning suggesting a SIM card to be installed if a UICC error is detected by the modem.
+  * Fixed bug causing cell network to be treated as offline if IPv4 not assigned.
```
Comment (b-gent) [inline doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:454]:
```suggestion
  * Fixed a bug causing the cell network to be treated as offline if IPv4 is not assigned.
```

**Review by eivindj-nordic — APPROVED** [review]

**Review by lemrey — APPROVED** [review]
Thanks

**Comment by glarsennordic** [comment]
I finally figured out a way to get my device onto IPv6 and confirmed this works

**Review by divipillai — APPROVED** [review]
